python/runtime/cudaq/distributed/mpi_comm_impl.cpp

```diff
@@ -0,0 +1,609 @@
+    return 1;
+  }
+}
+
+/// @brief Wrapper of MPI_Allreduce with MPI_IN_PLACE
+static int mpi_AllreduceInplace(const cudaqDistributedCommunicator_t *comm,
+                                void *recvBuffer, int32_t count,
+                                DataType dataType, ReduceOp opType) {
+  try {
+    auto mpiMod = py::module::import("mpi4py.MPI");
+    auto inplaceAddr = mpiMod.attr("IN_PLACE");
+    py::tuple recvBuf = py::make_tuple(packData(recvBuffer, count, dataType),
+                                       count, convertType(dataType));
+    auto pyComm = unpackMpiCommunicator(comm);
+
+    pyComm.attr("Allreduce")(inplaceAddr, recvBuf, convertType(opType));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+}
+
+/// @brief Wrapper of MPI_Allgather
+static int mpi_Allgather(const cudaqDistributedCommunicator_t *comm,
+                         const void *sendBuffer, void *recvBuffer,
+                         int32_t count, DataType dataType) {
+  try {
+    auto pyComm = unpackMpiCommunicator(comm);
+    const auto size = pyComm.attr("Get_size")().cast<int>();
+    pyComm.attr("Allgather")(packData(sendBuffer, count, dataType, true),
+                             packData(recvBuffer, size * count, dataType));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+}
+
+/// @brief Wrapper of MPI_Allgatherv
+static int mpi_AllgatherV(const cudaqDistributedCommunicator_t *comm,
+                          const void *sendBuf, int sendCount, void *recvBuf,
+                          const int *recvCounts, const int *displs,
+                          DataType dataType) {
+  try {
+    auto pyComm = unpackMpiCommunicator(comm);
+    const auto size = pyComm.attr("Get_size")().cast<int>();
+    py::array_t<int> pyRecvCounts(size);
+    py::array_t<int> pyDispls(size);
+    auto pyRecvCountsData = pyRecvCounts.mutable_unchecked<1>();
+    auto pyDisplsData = pyDispls.mutable_unchecked<1>();
+    for (int i = 0; i < size; ++i) {
+      pyRecvCountsData(i) = recvCounts[i];
+      pyDisplsData(i) = displs[i];
+    }
+    pyComm.attr("Allgatherv")(
+        packData(sendBuf, sendCount, dataType, true),
+        py::make_tuple(packData(recvBuf, size * sendCount, dataType),
+                       pyRecvCounts, pyDispls, convertType(dataType)));
+    return 0;
+  } catch (std::exception &e) {
+    std::cerr << "[mpi4py] Caught exception \"" << e.what() << "\"\n";
+    return 1;
+  }
+  return 0;
+}
+
+/// @brief Wrapper of MPI_Isend and track pending requests for synchronization
+static int mpi_SendAsync(const cudaqDistributedCommunicator_t *comm,
+                         const void *buf, int count, DataType dataType,
+                         int peer, int32_t tag) {
+  if (PendingRequest::g_requests[comm].nActiveRequests == 2)
```

Comment:
What is the reason for all of these `== 2` checks? Do we not support >2 active requests for some reason?